### PR TITLE
[cmake] add a cmake build script to build OpenThread

### PR DESCRIPTION
**Pull request by zhanglongxia**
This PR adds a bash script to compile the platform that supports cmake.

Usage Examples:
    ./script/cmake-build cc1352
    ./script/cmake-build posix -DOT_MAC_FILTER=on

### CODE REVIEW

File: script/cmake-build [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    fi
+
+    if [[ "${platform}" == "cc2650" ]]; then
+        targets+=("ot-cli-mtd" "ot-ncp-mtd")
+    fi
+
+    build "${platform}" "${options[*]}" "${targets[*]-}"
+}
+
+main "$@"
```
Comment (FiveDimensions) [inline script/cmake-build:83]: nit: 1. Is it possible to add default options for vendor platforms and posix, simulation? 
2. Add usage for the script?

Comment (simonlingoogle) [reply script/cmake-build:83]: 1) will be convinient for posix and platforms. Furthermore, can we provide default options for all platforms?

File: script/cmake-build [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    exit 1
+}
+
+build()
+{
+    platform=$1
+    options=($2)
+    targets=($3)
+
+    mkdir -p build/${platform}
```
Comment (simonlingoogle) [inline script/cmake-build:46]: what if the directory already exists?

File: script/cmake-build (hunk 1 above)
Comment (zhanglongxia) [reply script/cmake-build:83]: Will update the code.

File: script/cmake-build (hunk 2 above)
Comment (zhanglongxia) [reply script/cmake-build:46]: The option "-p" guarantees that if the directory already exits, it won't generate an error.

**Comment by zhanglongxia** [comment]
@bukepo Yes, we can let script/test leverage the new script.

File: examples/platforms/cc1352/arm-none-eabi.cmake [hunk 3]
```diff
@@ -28,6 +28,7 @@
 
 set(CMAKE_SYSTEM_NAME              Generic)
 set(CMAKE_SYSTEM_PROCESSOR         ARM)
+set(CMAKE_BUILD_TYPE               Release CACHE STRING "Set build type to release to save memory")
```
Comment (bukepo) [inline examples/platforms/cc1352/arm-none-eabi.cmake:31]: I think we should let user specify the build type.

File: script/check-arm-build-cmake [hunk 4]
```diff
@@ -29,174 +29,96 @@
 
 set -euxo pipefail
 
-readonly OT_SRCDIR="$(pwd)"
-readonly OT_BUILDDIR="$OT_SRCDIR/build"
+readonly OT_BUILDDIR="$(pwd)/build"
 
 readonly OT_COMMON_OPTIONS=(
-    "-DOT_COMPILE_WARNING_AS_ERROR=ON"
+    "-DOT_COMPILE_WARNING_AS_ERROR=on"
```
Comment (bukepo) [inline script/check-arm-build-cmake:35]: keep using ON/OFF?

File: script/check-arm-build-cmake [hunk 5]
```diff
@@ -29,174 +29,96 @@
...
+readonly OT_BASIC_CHECK_OPTIONS=(
+    "-DOT_COMMISSIONER=on"
+    "-DOT_DHCP6_CLIENT=on"
+    "-DOT_DHCP6_SERVER=on"
+    "-DOT_DNS_CLIENT=on"
+    "-DOT_JOINER=on"
+    "-DOT_SLAAC=on"
+)
+
+die()
```
Comment (bukepo) [inline script/check-arm-build-cmake:47]: `set -euxo pipefail` is set, so can we avoid using this?

File: script/check-arm-build-cmake [hunk 6]
```diff
@@ -29,174 +29,96 @@
...
-        "-DOT_COMMISSIONER=on"
-        "-DOT_DHCP6_CLIENT=on"
-        "-DOT_DHCP6_SERVER=on"
-        "-DOT_DNS_CLIENT=on"
-        "-DOT_JOINER=on"
-        "-DOT_SLAAC=on"
-        "-DCMAKE_BUILD_TYPE=Release"
-    )
+    local options=("${OT_COMMON_OPTIONS[@]}"
+        "${OT_BASIC_CHECK_OPTIONS[@]}")
```
Comment (bukepo) [inline script/check-arm-build-cmake:61]:
```suggestion
```

File: script/check-arm-build-cmake [hunk 7]
```diff
@@ -29,174 +29,96 @@
...
 
     reset_source
-    mkdir -p "$OT_BUILDDIR"
-    cd "$OT_BUILDDIR"
-    cmake -GNinja "${options[@]}" ..
-    ninja
-    cd "$OT_SRCDIR"
+    "$(dirname "$0")"/cmake-build cc1352 "${options[@]}"
```
Comment (bukepo) [inline script/check-arm-build-cmake:64]:
```suggestion
    "$(dirname "$0")"/cmake-build cc1352 "${OT_COMMON_OPTIONS[@]}" "${OT_BASIC_CHECK_OPTIONS[@]}"
```

File: script/test [hunk 8]
```diff
@@ -346,43 +285,20 @@ EXAMPLES:
...
-    builddir="${OT_BUILDDIR}/cmake/openthread-daemon"
-    (mkdir -p "${builddir}" \
-        && cd "${builddir}" \
-        && cmake -GNinja -DOT_PLATFORM="posix" \
-            -DOT_DAEMON=1 -DOT_PLATFORM_NETIF=1 -DOT_PLATFORM_UDP=1 "${options[@]}" "${ot_extra_options[@]}" "${OT_SRCDIR}" \
-        && ninja package \
-        && ls "${builddir}"/openthread-daemon-*.deb)
+    local options=("-DCMAKE_BUILD_TYPE=Release")
+
+    export OT_CMAKE_NINJA_TARGET="package"
```
Comment (bukepo) [inline script/test:290]: I tend not to export.

File: script/test [hunk 9]
```diff
@@ -93,75 +63,44 @@ build_simulation()
...
+        options+=("-DOT_CSL_RECEIVER=ON")
     fi
 
-    local builddir="${OT_BUILDDIR}/cmake/openthread-simulation-${version}"
-    (mkdir -p "${builddir}" \
-        && cd "${builddir}" \
-        && cmake -GNinja "${options[@]}" "${ot_extra_options[@]}" "${OT_SRCDIR}" \
-        && ninja)
+    export OT_CMAKE_BUILD_DIR="${OT_BUILDDIR}/cmake/openthread-simulation-${version}"
+    "$(dirname "$0")"/cmake-build simulation "${options[@]}" "${ot_extra_options[@]}"
```
Comment (bukepo) [inline script/test:70]:
```suggestion
    OT_CMAKE_BUILD_DIR="${OT_BUILDDIR}/cmake/openthread-simulation-${version}" "${OT_SRCDIR}"/cmake-build simulation "${options[@]}" "${ot_extra_options[@]}"
```

File: script/cmake-build [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+set -euo pipefail
+
+OT_CMAKE_NINJA_TARGET=${OT_CMAKE_NINJA_TARGET:-}
```
Comment (bukepo) [inline script/cmake-build:32]: how about make this the first argument of this script and use -DOT_PLATFORM= to specify the platform just like other options?

File: script/cmake-build [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    "-DOT_SERVICE=on"
+    "-DOT_SNTP_CLIENT=on"
+    "-DOT_COVERAGE=on"
+    "-DOT_LOG_LEVEL_DYNAMIC=on"
+    "-DOT_COMPILE_WARNING_AS_ERROR=on"
+)
+
+die()
+{
+    echo " ** ERROR: Openthread cmake doesn't support platform \"$1\""
```
Comment (FiveDimensions) [inline script/cmake-build:67]:
```suggestion
    echo " ** ERROR: OpenThread CMake doesn't support platform \"$1\""
```

File: script/cmake-build [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    if [[ $# == 0 ]]; then
+        print_usage
+    fi
+
+    local platform="$1"
+    shift
+    local local_options=()
+    local options=(
+        "-DOT_PLATFORM=${platform}"
+        "-DOT_SLAAC=on"
```
Comment (FiveDimensions) [inline script/cmake-build:123]: why only add this option?

File: examples/platforms/cc1352/arm-none-eabi.cmake (hunk 3 above)
Comment (FiveDimensions) [reply examples/platforms/cc1352/arm-none-eabi.cmake:31]: can we move it into script/cmake-build if  user does not specify CMAKE_BUILD_TYPE?

File: script/cmake-build [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+set -euo pipefail
```
Comment (jwhui) [inline script/cmake-build:30]: What is the purpose and scope of this script? Can we put some documentation here that provides more information? For example, does this script support use cases where the user wants a different build option value than whats included in this script for a specific platform?

Comment (zhanglongxia) [reply script/cmake-build:62]: Already added more information to describe the purpose and scope of this script.

File: script/cmake-build [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+#  Examples:
+#
+#  script/cmake-build simulation
+#
+#  script/cmake-build simulation -DOT_FULL_LOGS=ON -DOT_CHANNEL_MANAGER=OFF
+#
+#  OT_CMAKE_NINJA_TARGET="ot-cli-mtd" OT_CMAKE_BUILD_DIR="./build/temp" script/cmake-build simulation -DOT_FULL_LOGS=ON -DOT_CHANNEL_MANAGER=OFF
+#
+
+set -euo pipefail
```
Comment (jwhui) [inline script/cmake-build:62]:
```suggestion
set -euxo pipefail
```
For greater visibility into what the script is doing?

File: script/cmake-build [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+#
+#  This script calls cmake&ninja to compile OpenThread for the given platform.
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply script/cmake-build]: Updated

File: script/cmake-build [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+#
+#  Compile with default build options:
+#
+#  script/cmake-build ${platform}
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply script/cmake-build]: Updated

File: script/cmake-build (hunk 14 above)
Comment (zhanglongxia) [reply script/cmake-build]: Updated

**Pull request merged**